### ISSUE TITLE
Memory leak caused when casting dReal output to Python list

### ISSUE THREAD

**Issue by jsreid13** [dreal/dreal4#97]
dReal version: 4.18.07.1
OS version: Ubuntu 16.04.5 LTS
Compiler: g++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
Bazel version: 
```
Build label: 0.16.1
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Aug 13 13:43:36 2018 (1534167816)
Build timestamp: 1534167816
Build timestamp as int: 1534167816
```
Build commit: 639a5be Renamed file to replace existing Dockerfile
Python version: 3.5

Code:
```python
from dreal.symbolic import Variable, logical_and, sin, cos
from dreal.api import CheckSatisfiability, Minimize

x = Variable("x")
y = Variable("y")
z = Variable("z")

f_sat = logical_and(0 <= x, x <= 10,
                    0 <= y, y <= 10,
                    0 <= z, z <= 10,
                    sin(x) + cos(y) == z)

result = CheckSatisfiability(f_sat, 0.001)
print(list(result))
```
Originally I wanted to cast ```result``` to a dictionary since it seemed to be in that form but that gave the error 
```
TypeError: cannot convert dictionary update sequence element #0 to a sequence
```
Which means that the output is not separated into a list of lists where the inner lists are pairs of [variable: range of values], so I tried converting result to a list to see if that works, and found the script stalled and started eating up all my memory. I'm not very experienced at debugging memory leaks, but here is the bottom chunk of the output from running ```valgrind --leak-check=yes python3 dreal_test.py```  to hopefully help debug the cause, the full output is much longer:
```
==18909== 1,664 bytes in 3 blocks are possibly lost in loss record 2,117 of 2,212
==18909==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==18909==    by 0x5D3C48: PyObject_Malloc (in /usr/bin/python3.5)
==18909==    by 0x518C92: _PyObject_GC_NewVar (in /usr/bin/python3.5)
==18909==    by 0x4ED0AF: PyFrame_New (in /usr/bin/python3.5)
==18909==    by 0x53B73B: PyEval_EvalFrameEx (in /usr/bin/python3.5)
==18909==    by 0x53B7E3: PyEval_EvalFrameEx (in /usr/bin/python3.5)
==18909==    by 0x540198: ??? (in /usr/bin/python3.5)
==18909==    by 0x540E4E: PyEval_EvalCode (in /usr/bin/python3.5)
==18909==    by 0x54A6B7: ??? (in /usr/bin/python3.5)
==18909==    by 0x4E9C35: PyCFunction_Call (in /usr/bin/python3.5)
==18909==    by 0x53DBBA: PyEval_EvalFrameEx (in /usr/bin/python3.5)
==18909==    by 0x540198: ??? (in /usr/bin/python3.5)
==18909==
==18909== 2,840 bytes in 5 blocks are possibly lost in loss record 2,132 of 2,212
==18909==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==18909==    by 0x5D402C: ??? (in /usr/bin/python3.5)
==18909==    by 0x5D43BB: PyObject_Realloc (in /usr/bin/python3.5)
==18909==    by 0x516BFB: _PyObject_GC_Resize (in /usr/bin/python3.5)
==18909==    by 0x4ED0F1: PyFrame_New (in /usr/bin/python3.5)
==18909==    by 0x53B73B: PyEval_EvalFrameEx (in /usr/bin/python3.5)
==18909==    by 0x540F9A: PyEval_EvalCodeEx (in /usr/bin/python3.5)
==18909==    by 0x4EBD22: ??? (in /usr/bin/python3.5)
==18909==    by 0x5C1796: PyObject_Call (in /usr/bin/python3.5)
==18909==    by 0x5C2579: _PyObject_CallMethodIdObjArgs (in /usr/bin/python3.5)
==18909==    by 0x5260C7: PyImport_ImportModuleLevelObject (in /usr/bin/python3.5)
==18909==    by 0x549E77: ??? (in /usr/bin/python3.5)
==18909==
==18909== 3,408 bytes in 6 blocks are possibly lost in loss record 2,155 of 2,212
==18909==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==18909==    by 0x5D3C48: PyObject_Malloc (in /usr/bin/python3.5)
==18909==    by 0x518C92: _PyObject_GC_NewVar (in /usr/bin/python3.5)
==18909==    by 0x4ED0AF: PyFrame_New (in /usr/bin/python3.5)
==18909==    by 0x53B73B: PyEval_EvalFrameEx (in /usr/bin/python3.5)
==18909==    by 0x540F9A: PyEval_EvalCodeEx (in /usr/bin/python3.5)
==18909==    by 0x4EBD22: ??? (in /usr/bin/python3.5)
==18909==    by 0x5C1796: PyObject_Call (in /usr/bin/python3.5)
==18909==    by 0x5C2579: _PyObject_CallMethodIdObjArgs (in /usr/bin/python3.5)
==18909==    by 0x5260C7: PyImport_ImportModuleLevelObject (in /usr/bin/python3.5)
==18909==    by 0x549E77: ??? (in /usr/bin/python3.5)
==18909==    by 0x4E9BA6: PyCFunction_Call (in /usr/bin/python3.5)
==18909==
==18909== LEAK SUMMARY:
==18909==    definitely lost: 0 bytes in 0 blocks
==18909==    indirectly lost: 0 bytes in 0 blocks
==18909==      possibly lost: 27,608 bytes in 48 blocks
==18909==    still reachable: 24,335,109 bytes in 756,535 blocks
==18909==                       of which reachable via heuristic:
==18909==                         newarray           : 56 bytes in 1 blocks
==18909==         suppressed: 0 bytes in 0 blocks
==18909== Reachable blocks (those to which a pointer was found) are not shown.
==18909== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==18909==
==18909== For counts of detected and suppressed errors, rerun with: -v
==18909== Use --track-origins=yes to see where uninitialised values come from
==18909== ERROR SUMMARY: 116049 errors from 90 contexts (suppressed: 0 from 0)
```

**Comment by soonho-tri** [comment]
> Originally I wanted to cast result to a dictionary since it seemed to be in that form but that gave the error

1. FYI, `result` is a python object representing a `Box` object in C++ which works like a dictionary. For example, you can do `result[x]` to extract the interval assignment for `x`. Is there something that you want to do with `result` but couldn't do while Python's dictionary provide it? Please let me know.

2. Regarding the memory leak: I've found that it was not really a memory leak, but python keeps calling `__getitem(i)__` for `i = 0, 1, 2, ...` when it handles `list(result)`. I've pushed a fix 5ca2fa6b989e26a89e23323c50223faf814e89f1. If you use this commit, `list(result)` should not give you a non-termination but it gives you a list of intervals which I think not so much valuable since it lost keys.

**Comment by soonho-tri** [comment]
In C++, we provide two `operator[]` overloading for the `Box` class:
 - `Box[int i]`: returns the i-th interval.
 - `Box[Variable x]`: returns the interval for `x`.

When I wrote its Python binding, I added two `__getitem__` methods. It seems that python picks the first one `Box[int i]` and treats it as a sequence type instead of a dictionary-like type. I'll investigate and change the python APIs so that it is considered as a dictionary-like type.

**Comment by jsreid13** [comment]
Thanks for the quick fix, I added the code from 5ca2fa6 to my repo and rebuilt using Bazel and can confirm that it fixes the non-termination and now returns a list of Intervals. 

Regarding your first point, Box does seem to work like a dictionary I just never tried that because I thought it would be safer converting everything to Python types when working with them which is good to know moving forward. However, Box doesn't seem to have the methods .items() and .values() which I need for finding certain values in the dictionary, and Interval isn't able to be converted to a list so I'm not sure how to access values from it. So for now I just read the entire Box as a string and parse that to get the corresponding values.

Also I tested ```result[x]``` and it does return an Interval as does ```result[y]```, ```result[z]``` and ```result[2]```, however weirdly enough if you give it any random Variable it still returns the interval for x, so ```result[Variable('foo')]``` will return [1.24723, 1.24758]. Not sure if this means the ```[](const Box& self, const Variable& var) { return self[var]; })``` needs some check to see if the var is actually defined within the Box.
